### PR TITLE
Indicator for unsynced local changes

### LINKED ISSUE #3524: Indicator for unsynced local changes
Originally reported on Google Code with ID 2638

```
I would like an indicator that shows that there are unsynced local changes (learning
progress).
I learn on many devices (staniory (PC) and mobile) in parallel and commonly sync to
many times because i am unsure about the state of the devices. “Did i learn anything
on the tablet today? Don’t know, better sync once more to be safe.”

It could be something unobtrusive. Maybe an * added to the “You learned X cards in
Y” line at the bottom of the deck picker.

Is this feature available on similar apps (…)?

Not really in learning apps afaik. The syncing is one of the great pluses of AnkiDroid
compared to other apps.
But Emacs uses a * to indicate unsaved (rather than unsynced) changes, which i find
quite useful.
```

Reported by `ospalh` on 2015-06-10 09:46:18

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
Users currently do not know if they have pending changes. We add a badge to fix this.

API 23+ due to DrawableWrapper

## Fixes
Fixes #3524

## Approach
* Add a badge to the sync icon
  * If the user does not have an account (clicking sync takes them to logon).
  * If the changes require a full sync
  * If the changes require a regular sync

* On each database query, check if the query starts with "

This badge is not set when changing decks - as it would be more annoying than helpful (even though changing decks is internally a change when syncing).

## How Has This Been Tested?
On my phone - Statistics and Browse do not trigger the icon, reviewing and adding notes do.

image
image

## Learning
* Not listing false starts  - see the issue
* Android Drawables need a `.setBounds()` call if they're not
* `setBounds()` can be negative and still work
* UI is hard
* https://developer.android.com/reference/android/graphics/drawable/DrawableWrapper is API23+
* https://stackoverflow.com/questions/13057782/dynamic-drawable-icon-for-actionbar-menu-item-android-actionbarsherlock/13058113#13058113

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/web/DatabaseChangeDecorator.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+ This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License along with
+ this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki.web;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/web/DatabaseChangeDecorator.java:17]: doesn't seem "web"-y to me, seems more utils-y to me ? Shame all the boilerplate is necessary!

File: AnkiDroid/src/main/java/com/ichi2/anki/SyncStatus.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+ This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License along with
+ this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/SyncStatus.java:17]: also maybe Utils-y ?

File: AnkiDroid/src/main/java/com/ichi2/ui/BadgeDrawableBuilder.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+    public BadgeDrawableBuilder(@NonNull Resources resources) {
+        mResources = resources;
+    }
+
+
+    public static void removeBadge(MenuItem menuItem) {
+        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+            return;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/ui/BadgeDrawableBuilder.java:45]: I'd prefer if this sort of stuff was in the Compat infrastructure so we didn't spread the API-dependent conditionals across the codebase. Not sure how to do that well when there are related public methods like here though :thinking: might be this is as good as it gets and we already have to do a full code scan for API codes anyway, so maybe just think about it for a timebox of no more than two minutes, and leave as is if it doesn't seem easy to fit into Compat/CompatV16/CompatV23

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/ui/BadgeDrawableBuilder.java:45]: I couldn't and can't see a way to make this fit in - the conditionals are (nicely) hidden away in the Builder so they won't be seen by consumers, but it's still unpleasant.

**Comment by david-allison-1** [comment]
Pushed - I suspect we may get a user or two asking about disabling this, but YAGNI for now.

File: AnkiDroid/src/main/java/com/ichi2/utils/SyncStatus.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import androidx.annotation.NonNull;
+
+public enum SyncStatus {
+    INCONCLUSIVE,
+    NO_ACCOUNT,
+    NO_CHANGES,
+    HAS_CHANGES,
+    FULL_SYNC;
+
+    private static boolean sPauseCheckingDatabase = false;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/utils/SyncStatus.java:34]: Also flagging - static state is typically something to be avoided, but we'd need DI to solve it.

File: AnkiDroid/src/main/res/values/04-network.xml [hunk 5]
```diff
@@ -62,6 +62,8 @@
     <string name="sync_uploading_message">Uploading…</string>
     <string name="sync_downloading_message">Downloading…</string>
     <string name="sync_title">Synchronization</string>
+    <string name="sync_title_full_sync">Sync (full)</string>
```
Comment (david-allison-1) [inline AnkiDroid/src/main/res/values/04-network.xml:65]: @mikehardy Bikeshedding - but I'd actually like to move these out of Network, and use a "Sync" string for the DeckPicker, would that be OK with you?

File: AnkiDroid/src/main/java/com/ichi2/utils/SyncStatus.java (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/utils/SyncStatus.java:34]: I'm okay with it when needed

**Comment by mikehardy** [comment]
Combined feedback from my wife and I: very helpful feature but (always a but...) she said it looks a little "mustardy" with regard to color, and could be more orange, and the diameter could be reduced by perhaps 33-50%, badges are usually pretty subtle yet still effective, right now it's a a bit glaring due to size

**Comment by mikehardy** [comment]
We think the red "not configured at all" icon is a good size though. It should be pretty glaring, but will likely result in a call to disable it in the future (MYAreGNI haha)

**Pull request merged**